pkg/controller/servicebindingrequest/retriever.go

```diff
@@ -1,292 +1,139 @@
-		}
-		return fmt.Sprintf("%v", value), sectionMap, nil
-	}
-	attrs := strings.SplitN(key, ".", 2)
-	newSectionMap, exists := sectionMap.(map[string]interface{})[attrs[0]]
-	if !exists {
-		return "", newSectionMap, nil
+// createServiceIndexPath returns a string slice with fields representing a path to a resource in the
+// environment variable context. This function cleans fields that might contain invalid characters to
+// be used in Go template; for example, a Group might contain the "." character, which makes it
+// harder to refer using Go template direct accessors and is substituted by an underbar "_".
+func createServiceIndexPath(name string, gvk schema.GroupVersionKind) []string {
+	return []string{
+		gvk.Version,
+		strings.ReplaceAll(gvk.Group, ".", "_"),
+		gvk.Kind,
+		strings.ReplaceAll(name, "-", "_"),
 	}
-	return r.getNestedValue(attrs[1], newSectionMap.(map[string]interface{}))
-}
 
-// getCRKey retrieve key in section from CR object, part of the "plan" instance.
-func (r *Retriever) getCRKey(u *unstructured.Unstructured, section string, key string) (string, interface{}, error) {
-	obj := u.Object
-	objName := u.GetName()
-	log := r.logger.WithValues("CR.Name", objName, "CR.section", section, "CR.key", key)
-	log.Debug("Reading CR attributes...")
+}
 
-	sectionMap, exists := obj[section]
-	if !exists {
-		return "", sectionMap, fmt.Errorf("Can't find '%s' section in CR named '%s'", section, objName)
+func buildServiceEnvVars(svcCtx *ServiceContext, globalEnvVarPrefix string) (map[string]string, error) {
+	prefixes := []string{}
+	if len(globalEnvVarPrefix) > 0 {
+		prefixes = append(prefixes, globalEnvVarPrefix)
 	}
-
-	log.WithValues("SectionMap", sectionMap).Debug("Getting values from sectionmap")
-	v, _, err := r.getNestedValue(key, sectionMap)
-	for k, v := range sectionMap.(map[string]interface{}) {
-		if _, ok := r.cache[section]; !ok {
-			r.cache[section] = make(map[string]interface{})
-		}
-		r.cache[section].(map[string]interface{})[k] = v
+	if len(svcCtx.EnvVarPrefix) > 0 {
+		prefixes = append(prefixes, svcCtx.EnvVarPrefix)
 	}
-	return v, sectionMap, err
+	return envvars.Build(svcCtx.EnvVars, prefixes...)
 }
 
-// read attributes from CR, where place means which top level key name contains the "path" actual
-// value, and parsing x-descriptors in order to either directly read CR data, or read items from
-// a secret.
-func (r *Retriever) read(envVarPrefix *string, cr *unstructured.Unstructured, place, path string, xDescriptors []string) error {
-	log := r.logger.WithValues(
-		"CR.Section", place,
-		"CRDDescription.Path", path,
-		"CRDDescription.XDescriptors", xDescriptors,
-	)
-	log.Debug("Reading CRDDescription attributes...")
-
-	// holds the secret name and items
-	secrets := make(map[string][]string)
+// GetEnvVars returns the data read from related resources (see ReadBindableResourcesData and
+// ReadCRDDescriptionData).
+func (r *Retriever) GetEnvVars() (map[string][]byte, error) {
```

Comment:
This will require a refactoring in the `Secret` type we use to [persist the service binding request secret](https://github.com/isutton/service-binding-operator/blob/e3d57437db9e265f55e0907aa2205cf89d66b9d3/pkg/controller/servicebindingrequest/secret.go#L30-L30). I believe that can be further simplified.